### DISCUSSION TITLE
How to build bot into one .exe file for any computers?

### DISCUSSION THREAD

**Discussion by Amirhan-Taipovjan-Greatest-I** [reo7sp/tgbot-cpp#238] [Q&A] [answered]
Hello!
I wanted to know how can I assemble/build My Telegram Bot into just one .exe file without additional packets because I want to host My bot on My old computer where I think it's probably not so easy to install all dependencies, files and other stuff.

**Comment by we-lu** [comment]
Hi,
without additional packets? The library is built statically by default. Whether you need to export additional dlls you have to check. If you have Git BASH, you can look at it there with `ldd`.

**Reply by Amirhan-Taipovjan-Greatest-I** [reply]
> Hi, without additional packets? The library is built statically by default. Whether you need to export additional dlls you have to check. If you have Git BASH, you can look at it there with `ldd`.

Please can you explan with more details because I don't understand about Git BASH and `ldd`.
If possible, I want to have only one file... if no, then how can I make a 'archive' with My Bot program and additional packets?

**Reply by we-lu** [reply]
Have you downloaded Git for Windows? If so, you can enter the command `ldd YourTelegramBot.exe` in git bash. This exe is created automatically when you build your project. Then you will see a list of all the dll that the exe needs to run. In the column on the right you see the path. In order for your exe to also run on another Windows computer, I would now include all dll within your archive, which may not be available on your old computer. All dll that are in C:\Windows will most likely also be present on your old computer. So, I would copy all other dll into the archive with your exe. If you now execute the exe on your old PC, all dependencies should be found there.

**Reply by Amirhan-Taipovjan-Greatest-I** [reply]
> ldd YourTelegramBot.exe

![image](https://user-images.githubusercontent.com/51203385/198067293-d1aa0d25-4200-427f-ab80-e4605e3b4e26.png)
And what should I do next?

**Reply by we-lu** [reply]
> Then you will see a list of all the dll that the exe needs to run. In the column on the right you see the path. In order for your exe to also run on another Windows computer, I would now include all dll within your archive, which may not be available on your old computer. All dll that are in C:\Windows will most likely also be present on your old computer. So, I would copy all other dll into the archive with your exe.

Copy the following files to your old computer:

- testbox.exe
- libssl-3-x64.dll
- libcrypto-3-x64.dll

And try to run the exe

**Reply by Amirhan-Taipovjan-Greatest-I** [reply]
> > Then you will see a list of all the dll that the exe needs to run. In the column on the right you see the path. In order for your exe to also run on another Windows computer, I would now include all dll within your archive, which may not be available on your old computer. All dll that are in C:\Windows will most likely also be present on your old computer. So, I would copy all other dll into the archive with your exe.
> 
> Copy the following files to your old computer:
> 
> * testbox.exe
> * libssl-3-x64.dll
> * libcrypto-3-x64.dll
> 
> And try to run the exe

I keep seeing this (file not found):
![image](https://user-images.githubusercontent.com/51203385/198117821-89ac0a31-7308-449e-8172-ec401173b52d.png)

If it can be useful, then My both computers have Win10 Pro.

**Reply by we-lu** [reply]
That's weird. Is it fixed if you also copy the VCRUNTIME140_1D.dll or are there even more error messages appearing?

**Reply by Amirhan-Taipovjan-Greatest-I** [reply]
> That's weird. Is it fixed if you also copy the VCRUNTIME140_1D.dll or are there even more error messages appearing?

I've decided to copy these files and it actually worked!
![image](https://user-images.githubusercontent.com/51203385/199280036-39d74c77-58cc-4dcd-a003-f5c8ac8a1183.png)

**Comment by reo7sp** [comment, marked as answer]
Current Conan configuration builds tgbot-cpp itself statically by default, but that does not automatically guarantee a single self-contained EXE: libcurl/OpenSSL and the MSVC runtime must also be linked statically.

Build a Release configuration and inspect the result with:

```powershell
dumpbin /DEPENDENTS YourTelegramBot.exe
```

If non-system DLLs are listed, either build those dependencies statically with the same Conan profile or place the matching Release DLLs beside the EXE. Install the Microsoft Visual C++ Redistributable when the dynamic MSVC runtime is used.

Do not deploy a Debug build or copy `VCRUNTIME140_1D.dll`: the `D` suffix is the debug CRT and it is not redistributable. Also, one EXE cannot be guaranteed to run on literally any Windows computer; CPU architecture and supported Windows version must match.
